Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h

```diff
@@ -0,0 +1,177 @@
+#include <map>
+#include <string>
+#include <sstream>
+#include <memory>
+#include <iostream>
+
+#ifdef RDK_USE_BOOST_SERIALIZATION
+#include <RDGeneral/Invariant.h>
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/serialization/vector.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+#endif
+
+namespace RDKit {
+class ROMol;
+class ChemicalReaction;
+
+namespace ScaffoldNetwork {
+
+struct RDKIT_SCAFFOLDNETWORK_EXPORT ScaffoldNetworkParams {
+  bool includeGenericScaffolds =
+      true;  ///< include scaffolds with all atoms replaced by dummies
+  bool includeGenericBondScaffolds =
+      false;  ///< include scaffolds with all bonds replaced by single bonds
+  bool includeScaffoldsWithoutAttachments =
+      true;  ///< remove attachment points from scaffolds and include the result
+  bool includeScaffoldsWithAttachments =
+      true;  ///< Include the version of the scaffold with attachment points
+  bool keepOnlyFirstFragment =
+      true;  ///<  keep only the first fragment from the bond breaking rule
+  bool pruneBeforeFragmenting =
+      true;  ///<  Do a pruning/flattening step before starting fragmenting
+  bool flattenIsotopes = true;  ///< remove isotopes when flattening
+  bool flattenChirality =
+      true;  ///< remove chirality and bond stereo when flattening
+  bool flattenKeepLargest =
+      true;  ///< keep only the largest fragment when doing flattening
+
+  std::vector<std::shared_ptr<ChemicalReaction>>
+      bondBreakersRxns;  ///< the reaction(s) used to fragment. Should expect a
+                         ///< single reactant and produce two products
+  ScaffoldNetworkParams()
+      : ScaffoldNetworkParams{
+            {"[!#0;R:1]-!@[!#0:2]>>[*:1]-[#0].[#0]-[*:2]"}} {};
+  ScaffoldNetworkParams(const std::vector<std::string> &bondBreakersSmarts);
+};
+
+enum class RDKIT_SCAFFOLDNETWORK_EXPORT EdgeType {
+  Fragment = 1,     ///< molecule -> fragment
+  Generic = 2,      ///< molecule -> generic molecule (all atoms are dummies)
+  GenericBond = 3,  ///< molecule -> generic bond molecule (all bonds single)
+  RemoveAttachment = 4,  ///< molecule -> molecule with no attachment points
+  Initialize = 5         ///< molecule -> flattened molecule
+};
+
+struct RDKIT_SCAFFOLDNETWORK_EXPORT NetworkEdge {
+  size_t beginIdx;
+  size_t endIdx;
+  EdgeType type;
+  NetworkEdge() : beginIdx(0), endIdx(0), type(EdgeType::Initialize){};
+  NetworkEdge(size_t bi, size_t ei, EdgeType typ)
+      : beginIdx(bi), endIdx(ei), type(typ){};
+  bool operator==(const RDKit::ScaffoldNetwork::NetworkEdge &o) const {
+    return (beginIdx == o.beginIdx) && (endIdx == o.endIdx) && (type == o.type);
+  }
+  bool operator!=(const RDKit::ScaffoldNetwork::NetworkEdge &o) const {
+    return (beginIdx != o.beginIdx) || (endIdx != o.endIdx) || (type != o.type);
+  }
+#ifdef RDK_USE_BOOST_SERIALIZATION
```

Comment:
Your suggestion of moving the serialization to a seperate file worked well for the substructlibrary search worked well and reduced the code clutter of ifdefs

https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/SubstructLibrary/SubstructLibrarySerialization.h

Reply:
good point. I will take a look at that

Reply:
Ok, just spent some time with this and I think it won't significantly simplify the code. 
The SubstructLibrary is in general more complex to serialize and, since you used the "non-invasive" approach, the serialization can be quite separate from the class.
We could do the same thing for the ScaffoldNetwork, but that just adds code complexity (some extra methods or functions like `initFromString()`) in order to remove #ifdefs from the header file. Doesn't seem worth it to me.